### PR TITLE
Init commit

### PR DESCRIPTION
**Pull request by ScorpioCPH**
Hi, this PR implements the following features:

- Listen on Pod and check `SchedulerName` and `Status.Phase`.
- If check pass, use `NodeSelector` to filter node label which marked by `resource-controller`.
- Bind this Pod to this matched Node.

**Future works:**

- Find a proper way to check whether this Pod is scheduled.
- More scheduling algorithms.

@ddysher @NickrenREN PTAL.

### CODE REVIEW

File: cmd/resource-scheduler/main.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+var (
+	kubeconfig  string
+	masterURL   string
+	threadiness int = 1
+)
+
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "~/.kube/config",
```
Comment (NickrenREN) [inline cmd/resource-scheduler/main.go]: this is not a build-in scheduler ?

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+	glog.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (s *Scheduler) runWorker() {
+	for s.processNextWorkItem() {
```
Comment (NickrenREN) [inline pkg/scheduler/scheduler.go]: sleep some time ?

**Comment by NickrenREN** [comment]
This is designed to be a new scheduler ? I find it just filters labels, what about other restrictions(such as node capacity, affinity ...) ? Any plans ? or can we just add predicate functions for this ?

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+	}
+
+	glog.Info("Setting up event handlers")
+	// Set up an event handler for when Pod resources change. This
+	// handler will lookup the owner of the given Pod, and if it is
+	// owned by a Pod resource will enqueue that Pod resource for
+	// processing. This way, we don't need to implement custom logic for
+	// handling Pod resources. More info on this pattern:
+	// https://github.com/kubernetes/community/blob/8cafef897a22026d42f5e5bb3f104febe7e29830/contributors/devel/schedulers.md
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Comment (zjj2wry) [inline pkg/scheduler/scheduler.go]: i think we should use a informer only list watch "spec.schedulerName=resource-scheduler"
```go
func NewPodInformer(client clientset.Interface, resyncPeriod time.Duration, schedulerName string) coreinformers.PodInformer {
	selector := fields.ParseSelectorOrDie(
		"spec.schedulerName=" + schedulerName +
			",status.phase!=" + string(v1.PodSucceeded) +
			",status.phase!=" + string(v1.PodFailed))
	lw := cache.NewListWatchFromClient(client.CoreV1().RESTClient(), string(v1.ResourcePods), metav1.NamespaceAll, selector)
	return &podInformer{
		informer: cache.NewSharedIndexInformer(lw, &v1.Pod{}, resyncPeriod, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
	}
}
```

**Comment by zjj2wry** [comment]
@ddysher @ScorpioCPH 
how can i find some information about user requirement？

`nodeaffinity require` same with node selector, but more flexiable.

File: pkg/scheduler/scheduler.go (hunk 2 above)
Comment (ScorpioCPH) [reply pkg/scheduler/scheduler.go]: I think `workqueue.Get()` in `processNextWorkItem()` will block the thread so there is no need to sleep.

File: cmd/resource-scheduler/main.go (hunk 1 above)
Comment (ScorpioCPH) [reply cmd/resource-scheduler/main.go]: Yes, it is a custom-scheduler.

File: pkg/scheduler/scheduler.go (hunk 3 above)
Comment (ScorpioCPH) [reply pkg/scheduler/scheduler.go]: Thanks, I will have a try.

**Comment by ScorpioCPH** [comment]
@NickrenREN Yes, this is a very simple implementation now, `affinity` and `predicate functions` will come soon i think.

**Comment by ScorpioCPH** [comment]
@zjj2wry what do you mean `user requirement` ? container resources ?

**Comment by zjj2wry** [comment]
@ScorpioCPH i do not clear our target，GPU schedule?

**Comment by ScorpioCPH** [comment]
@zjj2wry FYI resource class design [proposal](https://github.com/caicloud/oscenter/pull/23)

#### Use Cases

- I have different types of CPUs (e.g. Intel E3/E5) or GPUs (e.g. Nvidia K80/1080Ti) on homogenous nodes, and i want to schedule pods on these different nodes.
- I want some nodes to be dedicated for special workflow.

**Comment by NickrenREN** [comment]
`this is a very simple implementation now, affinity and predicate functions will come soon i think.`
Not adding predicate functions for this new scheduler, what i am saying is: can we just add a new predicate function for `device plugins` in the existing scheduler instead of adding a new scheduler ?

**Comment by ScorpioCPH** [comment]
@NickrenREN Hi, this design is expected to no Kubernetes code changes :)

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+	glog.Infof("syncHandler, nodeSelector: %+v", pod.Spec.NodeSelector)
+	if len(pod.Spec.NodeSelector) == 0 {
+		glog.V(4).Infof("no nodeSelector specified, forget this pod '%s'", pod.Name)
+		return nil
+	}
+
+	// only support one NodeSelectorTerm now
+
+	selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
```
Comment (ddysher) [inline pkg/scheduler/scheduler.go]: where did you get this selector ?!  Shouldn't you extract resource class from Pod spec, and match it against node capacity and resource class CRD?

**Comment by ddysher** [comment]
@ScorpioCPH Please send a detailed design doc first. How scheduler works and how it works with resource controller (e.g.g what's the boundary)

**Comment by ScorpioCPH** [comment]
[design doc](https://docs.google.com/document/d/1EyOUHah_4hx3QCY0FGyM8n-1FHhIIlG47ws6BgDPfUk/edit?usp=sharing), PTAL.

**Pull request closed without merging**